Upgrade f-interpolator

som-snytt opened the pull request:
Fixes #9939
Fixes #11256
Fixes #11750
Fixes #13293

anatoliykmetyuk commented:
@dos65 @som-snytt any news on this one?

som-snytt commented:
It's not pretty but is mergeable. Future improvements would include optimizing and constant folding.

I can give it my eyes one more time this week.

anatoliykmetyuk commented:
So you're still planning to do some work on it? Or can @dos65  review it?

som-snytt commented:
I'll give it a quick pre-review and click ready for review.

My concern would be, it's only a couple hundred LOC but does the core project really want to ingest it? In scala 2 it was "just a macro", albeit "fast-tracked"; this solution for Scala 3 is more tightly coupled.

f-interpolator (as a demo feature for macros) feels more like a 3rd party lint or scalafix feature.

Alternatively, built-in interpolation could be re-defined as f-interpolation (that is, spec updated to say ordinary strings are always f-interpolated). In that case, investing in it makes more sense.

dos65 commented on compiler/src/dotty/tools/dotc/transform/localopt/StringContextChecker.scala:
```diff
@@ -49,7 +49,7 @@ object StringContextChecker {
       * @return true if an error/warning has been reported, false
       */
-    def hasReported() : Boolean
+    def hasReported : Boolean
```
This one might be reverted to minimize diff

som-snytt replied:
That was my favorite itch to scratch.

dos65 commented on compiler/src/dotty/tools/dotc/transform/localopt/StringInterpolatorOpt.scala:
```diff
@@ -115,53 +115,57 @@ class StringInterpolatorOpt extends MiniPhase {
-            if (!str.const.stringValue.isEmpty) concat(str)
-          }
-          result
-        case Apply(intp, args :: Nil) if sym.eq(defn.StringContext_f) =>
```
The same thing with minimizing diff here.
It seems that required changes might be applied only for body of this case-statement
`case Apply(intp, args :: Nil) if sym.eq(defn.StringContext_f) =>`

som-snytt replied:
This refactor was sequestered in the 2nd commit.

dos65 commented on compiler/src/dotty/tools/dotc/transform/localopt/FormatChecker.scala:
```diff
@@ -0,0 +1,247 @@
+import StringContextChecker.InterpolationReporter
+
+/** Formatter string checker. */
+abstract class FormatChecker(using reporter: InterpolationReporter):
```
It's needed either to move it's body into existing `StringContextChecker` or remove `StringContextChecker`

som-snytt replied:
Removed.

dos65 commented on compiler/src/dotty/tools/dotc/transform/localopt/FormatChecker.scala:
```diff
@@ -0,0 +1,247 @@
+
+  // Pick the first runtime type which the i'th arg can satisfy.
+  // If conversion is required, implementation must emit it.
+  def argType(argi: Int, types: ClassTag[?]*): ClassTag[?]
```
I noticed that `FormatChecker` is introduced as an abstract class and as I understood it was done to have an ability to have a unit test with specific implementation. It also required to use `ClassTag` instead of `Type` as it would require proper context initialization. Did I get it right?

What do you think about moving all logic into a single class?
That would allow to use `Type` in `Conversion`s and other places and the part with mapping `ClassTag` to `Type` won't be required. All logic will be located in a single file that is an additional bonus.

I understand that it will be hard to keep `FormatCheckerTest` but it seems that at this moment when we got parity with migrated tests from Scala2 compiler this unit test won't be so useful as at development moment.

som-snytt replied:
Actually, I want to backport to Scala 2, so a goal was for some target independence. That's useful for any future fixes or improvements.

`ClassTag` sufficed. I think I started by abstracting over `Type` but that was baroque.

I'll take another look at the class structure. I remember an open issue was optimization or at least constant folding; not sure if that is better solved here or generally. Other optimizations would be along the lines of "replace %d with .toInt" or whatever.

dos65 reviewed (commented):
@som-snytt Nice work! Glad to see that the parity with Scala2 is finally obtained.

I left some comments/suggestions

som-snytt commented:
Sorry I was busy this past week, I meant to update this over the weekend! Still plan to look at this next time I touch the keyboard. Thanks for giving it eyes!

som-snytt commented:
I guess the best part is 850 lines added, 851 lines removed.

anatoliykmetyuk commented:
@dos65 @som-snytt what's the status on this one? Is the PR review-ready in which case WDYT about it @dos65 ?

dos65 commented:
@anatoliykmetyuk I think it's almost ready to merge. The only thing I'm worried that the [`FormatChecker` implementation might be simplified](https://github.com/lampepfl/dotty/pull/13367#discussion_r750461063) if it conversions would be based on `Type` instead of `ClassTag` + from general approach I see that `ClassTag` isn't used in this way in scala3-compiler. 

@som-snytt Don't you mind if I'll try to address this thing and will do a PR onto your branch? I think I will have time for that on this week.

som-snytt commented:
My previous comment on use of `ClassTag` still seems true to me:

> Actually, I want to backport to Scala 2, so a goal was for some target independence. That's useful for any future fixes or improvements.

> ClassTag sufficed. I think I started by abstracting over Type but that was baroque.

I didn't follow up on ensuring `s"hello, world"` is a constant string, or `s"hello, ${ "world" }"` etc:

> I'll take another look at the class structure. I remember an open issue was optimization or at least constant folding; not sure if that is better solved here or generally. Other optimizations would be along the lines of "replace %d with .toInt" or whatever.

dos65 commented:
@som-snytt 

> Actually, I want to backport to Scala 2, so a goal was for some target independence.

What do you mean by backport to Scala2? Do you want to port it back? I thought that this PR is a port from Scala2 :thinking: 

> I didn't follow up on ensuring s"hello, world" is a constant string, or s"hello, ${ "world" }" etc:

I've checked this sample, it's constant. Is there anything else to check?

I think there is a need to ask someone from core compiler team for a final review as it's a large PR.
From my side, I see that it works well, implementation is close to scala2-compiler, all tests from Scala2 were ported and pass :tada:

som-snytt commented:
yes, I intended to backport the forwardport. 😄 

I won't stand in the way of whatever you'd like to improve, for some definition of improve.

Mostly I feel that the compiler should not ingest all this code that ought to be library code, but the first step is just correctness. There is no binary compat constraint so it can be fixed/improved at any time.

som-snytt commented:
While I think of it, `config.Settings` is an example use of `ClassTag` to represent erased types in compiler.

som-snytt commented:
Embraced @dos65 's suggestions and dropped some code. Although I embraced the suggestions, I dropped some braces. I remember how nice it was to have the braces removed from my teeth in middle school.

anatoliykmetyuk commented:
Sounds like this one is ready to be merged and @dos65 doesn't have any more suggestions? We'll look into who can review it in the compiler team. Thanks @som-snytt for all the work!

som-snytt commented on compiler/src/dotty/tools/dotc/transform/localopt/FormatChecker.scala:
```diff
@@ -0,0 +1,283 @@
+    types.find(t => argConformsTo(argi, tpe, t))
+      .orElse(types.find(t => argConvertsTo(argi, tpe, t)))
+      .getOrElse {
+        report.argError(s"Found: ${tpe.show}, Required: ${types.mkString(", ")}", argi)
```
not pretty-printed. Add a test. Maybe there was a partest.

som-snytt commented on compiler/src/dotty/tools/dotc/transform/localopt/FormatChecker.scala:
```diff
@@ -0,0 +1,283 @@
+    // is the specifier OK with the given arg
+    def accepts(arg: Type): Boolean =
+      kind match
+        case BooleanXn  => arg == defn.BooleanType orElse warningAt(CC)("Boolean format is null test for non-Boolean")
```
Probably the reason was to enable implicit conversion to boolean. If `Any` is a variant, then no conversion is required or summoned.

som-snytt commented:
Improved test output to show where `// error` is extra or missing. A test had 50 errors, impossible to correlate while sober.

anatoliykmetyuk commented:
@som-snytt please let us know once this is review ready :)

som-snytt commented:
Definitely ready. I could definitely make improvements indefinitely.

som-snytt commented:
Rebased on phase name conflict.

Tweaked the f-usages that required backslashes to obviate returning to it later.

An unbootstrapped compiler will show doubled backslash in a few messages:
```
<   |              octal escape literals are unsupported: use \u0000 instead
---
>   |              octal escape literals are unsupported: use \\u0000 instead
```
"Caveat backslasher."

anatoliykmetyuk reviewed (approved):
LGTM🎉
Thanks @som-snytt for all the work, and @dos65 for reviewing it!

mbovel commented:
Non-bootstrapped tests are failing. See [logs](https://github.com/lampepfl/dotty/runs/5111457253?check_suite_focus=true).

mbovel commented:
Could reproduce locally:

```
-- Error: tests/neg/t7292-removal.scala:2:14 ---------------------------------------------------------------------------
2 |  val chr1 = '\0' // error
              ^
              octal escape literals are unsupported: use \\u0000 instead
-- Error: tests/neg/t7292-removal.scala:3:17 ---------------------------------------------------------------------------
3 |  val str1 = "abc\123456" // error
                 ^
                 octal escape literals are unsupported: use \\u0053 instead
-- Error: tests/neg/t7292-removal.scala:4:12 ---------------------------------------------------------------------------
4 |  val lf = '\012' // error
            ^
            octal escape literals are unsupported: use \n instead
```

The problem seems to be the duplicated slash in ` \\u`.
